dataframely/columns/array.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (c) QuantCo 2025-2025
+# SPDX-License-Identifier: BSD-3-Clause
+
+from __future__ import annotations
+
+from collections.abc import Callable, Sequence
+from typing import Any, Literal, cast
+
+import polars as pl
+
+from dataframely._compat import pa, sa, sa_TypeEngine
+from dataframely.random import Generator
+
+from ._base import Column
+from .struct import Struct
+
+
+class Array(Column):
+    """A fixed-shape array column."""
+
+    def __init__(
+        self,
+        inner: Column,
+        shape: int | tuple[int, ...],
+        *,
+        nullable: bool = True,
+        # polars doesn't yet support grouping by arrays,
+        # see https://github.com/pola-rs/polars/blob/810df456/crates/polars-core/src/frame/group_by/into_groups.rs#L365
+        primary_key: Literal[False] = False,
+        check: Callable[[pl.Expr], pl.Expr] | None = None,
+        alias: str | None = None,
+        metadata: dict[str, Any] | None = None,
+    ):
+        """
+        Args:
+            inner: The inner column type. No validation rules on the inner type are supported yet.
+            shape: The shape of the array.
+            nullable: Whether this column may contain null values.
+            primary_key: Whether this column is part of the primary key of the schema.
+                Not yet supported for the Array type.
+            check: A custom check to run for this column. Must return a non-aggregated
+                boolean expression.
+            alias: An overwrite for this column's name which allows for using a column
+                name that is not a valid Python identifier. Especially note that setting
+                this option does _not_ allow to refer to the column with two different
+                names, the specified alias is the only valid name.
+            metadata: A dictionary of metadata to attach to the column.
+        """
+        if primary_key:
+            raise ValueError(
+                "`primary_key=True` is not yet supported for the Array type."
+            )
```

Comment:
I don't think we need to cover this case specifically if the type annotation already prevents passing in `True`.

Reply:
my idea was that the following situation might be very common:
- a library user being used to the primary_key argument accepting any boolean value, and
- the same user not using typing checks
in those cases, it is easy to just write `primary_key=True`, the error would show up later and the message from polars would be more difficult to decipher. Hence the proactive exception.

Let me know if you still think this is unnecessary.

Reply:
> the same user not using typing checks

So far, we've ignored this setting as our stance was that typing is so integral to the use of dataframely that it doesn't make sense to use it without. This choice is mirrored throughout the codebase (which has also prevented writing tests that require ignoring the type checker).

For consistency, I'd propose to get rid of it. If we ever want to make dataframely nicely usable without typing, we'd have to touch a couple places, I think.